src/libraries/System.Private.CoreLib/src/System/IO/StringReader.cs

```diff
@@ -175,38 +179,47 @@ public override string ReadToEnd()
         //
         public override string? ReadLine()
         {
-            if (_s == null)
+            string? s = _s;
+            if (s == null)
             {
-                throw new ObjectDisposedException(null, SR.ObjectDisposed_ReaderClosed);
+                ThrowHelper.ThrowObjectDisposedException_ReaderClosed();
             }
+            var pos = _pos;
+            if ((uint)pos >= (uint)s.Length)
+                return null;
 
-            int i = _pos;
-            while (i < _length)
+            ReadOnlySpan<char> remaining = s.AsSpan(pos);
+            int foundLineLength = remaining.IndexOfAny('\r', '\n');
+            if (foundLineLength >= 0)
             {
-                char ch = _s[i];
-                if (ch == '\r' || ch == '\n')
+                string result = foundLineLength == 0
```

Comment:
In general not calling `Substring` since this is defined with a lot of checks (that are unnecessary in the context), which explains why things have been changed to what they are. To stay in `Span`-land.

https://github.com/dotnet/runtime/blob/12a8819eee9865eb38bca6c05fdece1053102854/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs#L1847-L1884

Reply:
Then this can be addressed in `string.Substring` in a future PR. We shouldn't add complexity to the call site to work around a perceived deficiency in the implementation of `Substring`.

Reply:
@GrabYourPitchforks I'm fine with changing this back, but some of this of course relates to `Substring` API and this not "knowing" constraints of input, while when span is created, those constraints are in place. So it's not necessarily because something is "wrong" with `Substring` but more that inside this context we can side step these checks by using already verified state. If that makes sense. Again this only matters - no big impact though - for very short lines, but this is where we have seen the regressions.

Reply:
I understand the desire to avoid the overhead. However, even this code has several different checks: the call to `Slice` incurs one, the `string` ctors incurs one, and the comparison `(foundLineLength == 0)` incurs one. So the suggestion is to remove all of the checks here let `Substring` handle them on your behalf. An optimized implementation of `Substring` (which, again, doesn't have to come with this PR) can give you near-equivalent performance, _and_ it would simplify this call site. That's a win-win.

Reply:
Alright 😊 I'll change it as soon as I have an available moment.